File: pkg/layers/memory.go
package layers

import (
	"math"
	"ratatoskr/pkg/client"
	"ratatoskr/pkg/repos"
	"ratatoskr/pkg/types"
	"ratatoskr/pkg/utils"
	"sort"
)

type MemoryLayer struct {
	child  Layer
	repo   *repos.MessageRepo
	logger *utils.Logger
}

func NewMemoryLayer(repo *repos.MessageRepo, child Layer) *MemoryLayer {
	logger := utils.NewLogger("Memory Layer")
	return &MemoryLayer{
		child,
		repo,
		logger,
	}
}

func cosineSimilarity(a, b []float32) float32 {
	var dotProduct, magnitudeA, magnitudeB float32
	for i := range a {
		dotProduct += a[i] * b[i]
		magnitudeA += a[i] * a[i]
		magnitudeB += b[i] * b[i]
	}
	return dotProduct / float32(math.Sqrt(float64(magnitudeA))*math.Sqrt(float64(magnitudeB)))
}

func (m *MemoryLayer) getContext(username string, inputMessage string) ([]types.StoredMessage, error) {
	var history []types.StoredMessage
	// Get the context in order to feed future prompts
	todaysMessages, err := m.repo.GetMessages(username)
	if err != nil {
		m.logger.Error("getContext, GetMessages", err)
		return nil, err
	}

	err, contextualMessages := m.repo.GetAllMessagesForUser(username)
	// take the top three embeddings that have cosine similarity
	msgEmbedding := client.Embed(inputMessage)
	for i := range contextualMessages {
		rank := cosineSimilarity(contextualMessages[i].Embedding, msgEmbedding)
		m.logger.Info("Calculated rank at %v\n", rank)
		contextualMessages[i].Rank = rank
	}

	// sort contextualmessages ranked highest to lowest
	sort.Slice(contextualMessages, func(i, j int) bool {
		return contextualMessages[i].Rank > contextualMessages[j].Rank
	})

	//append the top messages to the history
	for i := 0; i < 10; i++ {
		if i < len(contextualMessages) {
			history = append(history, contextualMessages[i])
		}
	}

	// Select only the latest 10 messages
	if len(history) > 10 {
		todaysMessages = todaysMessages[len(todaysMessages)-10:]
		history = append(history, todaysMessages...)
	}

	for _, ctxMsg := range history {
		m.logger.Info("Context message: %v: %v\n", ctxMsg.Role, ctxMsg.Message)
	}

	return history, nil
}

func (m *MemoryLayer) PassThrough(req *types.RequestMessage) (types.ResponseMessage, error) {
	inputMessage := repos.NewStoredMessage(repos.User, req.Message)
	m.repo.SaveMessage(req.UserName, *inputMessage)

	history, err := m.getContext(req.UserName, req.Message)
	if err != nil {
		m.logger.Error("PassThrough, getContext", err)
		return types.ResponseMessage{}, err
	}
	req.Context = history

	// Now pass through child layer
	res, err := m.child.PassThrough(req)
	if err != nil {
		m.logger.Error("PassThrough, child layer", err)
		return types.ResponseMessage{}, err
	}

	outputMessage := repos.NewStoredMessage(repos.Assistant, res.Message)
	m.repo.SaveMessage(req.UserName, *outputMessage)

	return res, nil
}

File: src/layers/memory.rs
use std::borrow::Borrow;

use serde::{Deserialize, Serialize};
use tracing::error;

use async_trait::async_trait;

use crate::{
    clients::{
        chatgpt::Role,
        muninn::{MunninClient, MunninClientImpl},
    },
    message_types::ResponseMessage,
    RequestMessage,
};

use super::Layer;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct StoredMessage {
    pub username: String,
    pub text: String,
    pub role: Role,
}

pub struct MemoryLayer {
    next: Box<dyn Layer>,
}

#[async_trait]
impl Layer for MemoryLayer {
    async fn execute(&mut self, message: &mut RequestMessage) -> ResponseMessage {
        let munnin_client = MunninClientImpl::new();

        let context = match munnin_client.get_context(message.username.clone()).await {
            Ok(context) => context,
            Err(err) => {
                error!("Failed to get context: {:?}", err);
                return ResponseMessage {
                    bytes: None,
                    options: None,
                    text: "Failed to get context".to_string(),
                };
            }
        };

        // Convert context to stored messages
        let mut stored_context: Vec<StoredMessage> = Vec::new();
        for chat_response in context {
            stored_context.push(StoredMessage {
                username: message.username.clone(),
                text: chat_response.content,
                role: match chat_response.role.as_str() {
                    "user" => Role::User,
                    "assistant" => Role::Assistant,
                    _ => Role::System,
                },
            });
        }

        message.context = stored_context;
        let res = self.next.execute(message).await;

        munnin_client
            .save(
                message.username.to_string().borrow(),
                "user".to_string(),
                message.text.clone(),
            )
            .await
            .unwrap();

        munnin_client
            .save(
                message.username.to_string().borrow(),
                "assistant".to_string(),
                res.text.clone(),
            )
            .await
            .unwrap();

        res
    }
}

impl MemoryLayer {
    pub fn new(next: Box<dyn Layer>) -> Self {
        MemoryLayer { next }
    }
}
